app/services/NpsService.scala

```diff
@@ -57,24 +49,18 @@ class NpsService @Inject()(repository: Repository, innerService: DirectNpsServic
           repository
             .set(nino, taxYear, data)
             .map(_ => Right(data))
-            .recover { case _ => Left(HttpResponse(INTERNAL_SERVER_ERROR)) }
       }
 }
 
 class DirectNpsService @Inject()(applicationConfig: ApplicationConfig, npsConnector: NpsConnector) {
   def getPayeATSData(nino: String, taxYear: Int)(
-    implicit hc: HeaderCarrier): Future[Either[HttpResponse, PayeAtsMiddleTier]] = {
-    val logger = Logger(getClass.getName)
-    npsConnector.connectToPayeTaxSummary(nino, taxYear - 1) map { response =>
-      response status match {
-        case OK => Right(response.json.as[PayeAtsData].transformToPayeMiddleTier(applicationConfig, nino, taxYear))
-        case _  => Left(response)
-      }
-    } recover {
-      case e: JsResultException => {
-        logger.error(s"Exception in NpsService parsing Json: $e", e)
-        Left(HttpResponse(INTERNAL_SERVER_ERROR))
-      }
+    implicit hc: HeaderCarrier): Future[Either[ServiceError, PayeAtsMiddleTier]] =
+    npsConnector.connectToPayeTaxSummary(nino, taxYear - 1) map {
+      case Right(value) =>
+        Right(value.json.as[PayeAtsData].transformToPayeMiddleTier(applicationConfig, nino, taxYear))
+      case Left(error) if error.statusCode == NOT_FOUND             => Left(NotFoundError(error.message))
```

Comment:
If passing around `UpstreamErrorRepsonse` is not possible then this logic should sit in the `apply` function of the `ServiceError` object

Reply:
done. Changed to UptreamErrorResponse